File: app/routers/jobs.py
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from app.middleware.auth_middleware import get_current_user, CurrentUser
from app.models.scraped_job import (
    JobSearchRequest,
    JSearchRequest,
    ScrapedJobListResponse,
    JSearchFetchResponse,
    FetchStatusResponse,
)
from app.services.jobs_service import (
    search_and_store_jobs,
    search_and_store_jsearch_jobs,
    list_scraped_jobs,
    get_fetch_status,
    get_job_counts,
)

from app.models.scraped_job import (
    JobSearchRequest,
    JSearchRequest,
    ScrapedJobListResponse,
    JSearchFetchResponse,
    FetchStatusResponse,
    JSearchSchedulerConfig,
    JSearchSchedulerStatusResponse,
)
from app.services.jsearch_scheduler_service import (
    start_jsearch_scheduler_for_org,
    stop_jsearch_scheduler_for_org,
    get_jsearch_scheduler_status_for_org,
)

from app.services.jobs_service import get_all_public_jobs

router = APIRouter(prefix="/api/v1/marketplace/jobs", tags=["Marketplace Jobs"])

@router.get("/public", response_model=ScrapedJobListResponse)
async def get_public_jobs():
    return await get_all_public_jobs()

@router.post("/search", response_model=ScrapedJobListResponse)
async def search_jobs(
    payload: JobSearchRequest,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return await search_and_store_jobs(payload, org_id=current_user.org_id)
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Scraping error: {exc}") from exc


@router.post("/jsearch", response_model=JSearchFetchResponse)
async def search_jobs_jsearch(
    payload: JSearchRequest,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return await search_and_store_jsearch_jobs(payload, org_id=current_user.org_id)
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"JSearch error: {exc}") from exc


@router.get("/fetch-status", response_model=List[FetchStatusResponse])
async def fetch_status(
    sites: str = Query("indeed,linkedin,glassdoor,zip_recruiter,google,jsearch"),
    current_user: CurrentUser = Depends(get_current_user),
):
    site_list = [s.strip() for s in sites.split(",") if s.strip()]
    return await get_fetch_status(org_id=current_user.org_id, sites=site_list)


@router.get("/counts")
async def job_counts(
    keyword: Optional[str] = Query(None),
    location: Optional[str] = Query(None),
    is_remote: Optional[str] = Query(None),
    date_posted: Optional[str] = Query(None),
    current_user: CurrentUser = Depends(get_current_user),
):
    return await get_job_counts(
        org_id=current_user.org_id,
        keyword=keyword,
        location=location,
        is_remote=is_remote,
        date_posted=date_posted,
    )


@router.get("/", response_model=ScrapedJobListResponse)
async def get_scraped_jobs(
    limit: int = Query(50, ge=1, le=100),
    skip: int = Query(0, ge=0),
    keyword: Optional[str] = Query(None),
    location: Optional[str] = Query(None),
    job_type: Optional[str] = Query(None),
    site: Optional[str] = Query(None),
    skills: Optional[str] = Query(None),
    current_user: CurrentUser = Depends(get_current_user),
):
    return await list_scraped_jobs(
        org_id=current_user.org_id,
        limit=limit,
        skip=skip,
        keyword=keyword,
        location=location,
        job_type=job_type,
        site=site,
        skills=skills,
    )

@router.get("/jsearch/scheduler/status", response_model=JSearchSchedulerStatusResponse)
async def jsearch_scheduler_status(
    current_user: CurrentUser = Depends(get_current_user),
):
    return await get_jsearch_scheduler_status_for_org(org_id=current_user.org_id)


@router.post("/jsearch/scheduler/start", response_model=JSearchSchedulerStatusResponse)
async def jsearch_scheduler_start(
    payload: JSearchSchedulerConfig,
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return await start_jsearch_scheduler_for_org(
            org_id=current_user.org_id,
            config=payload,
        )
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Scheduler start error: {exc}") from exc


@router.post("/jsearch/scheduler/stop", response_model=JSearchSchedulerStatusResponse)
async def jsearch_scheduler_stop(
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return await stop_jsearch_scheduler_for_org(org_id=current_user.org_id)
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Scheduler stop error: {exc}") from exc

File: node-server/src/routes/jobs.js
import { Router } from 'express';
import { authenticate } from '../middleware/authMiddleware.js';
import {
  getPublicJobs,
  getPublicJobsCount,
  getFetchStatus,
  getJobCounts,
  listScrapedJobs,
  getPublicFilterCounts,
  searchAndStoreJsearchJobs,
} from '../services/jobsService.js';
import {
  startJsearchSchedulerForOrg,
  stopJsearchSchedulerForOrg,
  getJsearchSchedulerStatusForOrg,
} from '../services/jsearchSchedulerService.js';
import {
  jsearchJobDetails,
  jsearchEstimatedSalary,
  jsearchCompanySalary,
} from '../services/jsearchService.js';

const router = Router();

/**
 * PUBLIC LIST ROUTE
 * Frontend currently calls:
 *   GET /api/v1/marketplace/jobs?limit=100&skip=0
 *   GET /api/v1/marketplace/jobs?site=indeed&limit=1
 *   GET /api/v1/marketplace/jobs?job_type=fulltime&site=jsearch&limit=1
 */
router.get('/', async (req, res) => {
  try {
    const {
      limit = 50,
      skip = 0,
      keyword,
      location,
      job_type,
      site,
      skills,
      fetch_all = 'false',
      date_posted = null,
    } = req.query;

    const result = await listScrapedJobs({
      orgId: null,
      limit: parseInt(limit, 10) || 50,
      skip: parseInt(skip, 10) || 0,
      keyword,
      location,
      jobType: job_type,
      site,
      skills,
      fetchAll: String(fetch_all).toLowerCase() === 'true',
      datePosted: date_posted,
    });

    res.json(result);
  } catch (err) {
    console.error('[GET /jobs] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to list jobs' });
  }
});

/**
 * PUBLIC COUNTS/FILTERS ROUTE
 * Useful for frontend filter summaries
 */
router.get('/counts', async (req, res) => {
  try {
    const {
      keyword,
      location,
      job_type,
      site,
      skills,
      date_posted = null,
    } = req.query;

    const result = await getPublicFilterCounts({
      keyword,
      location,
      jobType: job_type,
      site,
      skills,
      datePosted: date_posted,
    });

    res.json(result);
  } catch (err) {
    console.error('[GET /jobs/counts] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get counts' });
  }
});

/**
 * LEGACY PUBLIC ROUTES
 */
router.get('/public', async (req, res) => {
  try {
    const { keyword, location, job_type, site, skills, page = 1 } = req.query;
    const result = await getPublicJobs({
      keyword,
      location,
      jobType: job_type,
      site,
      skills,
      page: parseInt(page, 10) || 1,
      limit: 200,
    });
    res.json(result);
  } catch (err) {
    console.error('[GET /jobs/public] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get public jobs' });
  }
});

router.get('/public/counts', async (req, res) => {
  try {
    const { keyword, location, job_type, site, skills } = req.query;
    const result = await getPublicJobsCount({
      keyword,
      location,
      jobType: job_type,
      site,
      skills,
    });
    res.json(result);
  } catch (err) {
    console.error('[GET /jobs/public/counts] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get public counts' });
  }
});

/**
 * AUTH ROUTES
 */
router.post('/jsearch', authenticate, async (req, res) => {
  try {
    const result = await searchAndStoreJsearchJobs({
      keywords: req.body.keywords,
      location: req.body.location,
      country: req.body.country || 'us',
      language: req.body.language || null,
      numPages: req.body.num_pages || 1,
      datePosted: req.body.date_posted || 'all',
      workFromHome: req.body.work_from_home || false,
      employmentTypes: req.body.employment_types || null,
      jobRequirements: req.body.job_requirements || null,
      radius: req.body.radius || null,
      excludeJobPublishers: req.body.exclude_job_publishers || null,
      useCursor: req.body.use_cursor || false,
      cursor: req.body.cursor || null,
    }, req.currentUser.orgId);

    res.json(result);
  } catch (err) {
    console.error('[POST /jobs/jsearch] error:', err);
    res.status(502).json({ detail: `JSearch error: ${err.message}` });
  }
});

router.get('/jsearch/job-details', authenticate, async (req, res) => {
  try {
    const { job_id, country = 'us', language } = req.query;
    if (!job_id) return res.status(400).json({ detail: 'job_id required' });
    res.json({ data: await jsearchJobDetails({ jobId: job_id, country, language }) });
  } catch (err) {
    console.error('[GET /jobs/jsearch/job-details] error:', err);
    res.status(502).json({ detail: `JSearch error: ${err.message}` });
  }
});

router.get('/jsearch/estimated-salary', authenticate, async (req, res) => {
  try {
    const { job_title, location, location_type = 'ANY', years_of_experience = 'ALL' } = req.query;
    if (!job_title || !location) {
      return res.status(400).json({ detail: 'job_title and location required' });
    }

    res.json({
      data: await jsearchEstimatedSalary({
        jobTitle: job_title,
        location,
        locationType: location_type,
        yearsOfExperience: years_of_experience,
      }),
    });
  } catch (err) {
    console.error('[GET /jobs/jsearch/estimated-salary] error:', err);
    res.status(502).json({ detail: `JSearch error: ${err.message}` });
  }
});

router.get('/jsearch/company-salary', authenticate, async (req, res) => {
  try {
    const { company, job_title, location, location_type = 'ANY', years_of_experience = 'ALL' } = req.query;
    if (!company || !job_title) {
      return res.status(400).json({ detail: 'company and job_title required' });
    }

    res.json({
      data: await jsearchCompanySalary({
        company,
        jobTitle: job_title,
        location,
        locationType: location_type,
        yearsOfExperience: years_of_experience,
      }),
    });
  } catch (err) {
    console.error('[GET /jobs/jsearch/company-salary] error:', err);
    res.status(502).json({ detail: `JSearch error: ${err.message}` });
  }
});

router.get('/fetch-status', authenticate, async (req, res) => {
  try {
    const sites = (req.query.sites || 'indeed,linkedin,glassdoor,zip_recruiter,google,jsearch')
      .split(',')
      .map(s => s.trim())
      .filter(Boolean);

    res.json(await getFetchStatus(req.currentUser.orgId, sites));
  } catch (err) {
    console.error('[GET /jobs/fetch-status] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get fetch status' });
  }
});

router.get('/org/counts', authenticate, async (req, res) => {
  try {
    const { keyword, location, is_remote, date_posted } = req.query;
    res.json(await getJobCounts({
      orgId: req.currentUser.orgId,
      keyword,
      location,
      isRemote: is_remote,
      datePosted: date_posted,
    }));
  } catch (err) {
    console.error('[GET /jobs/org/counts] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get org job counts' });
  }
});

router.get('/jsearch/scheduler/status', authenticate, async (req, res) => {
  try {
    res.json(await getJsearchSchedulerStatusForOrg(req.currentUser.orgId));
  } catch (err) {
    console.error('[GET /jobs/jsearch/scheduler/status] error:', err);
    res.status(500).json({ detail: err.message || 'Failed to get scheduler status' });
  }
});

router.post('/jsearch/scheduler/start', authenticate, async (req, res) => {
  try {
    res.json(await startJsearchSchedulerForOrg(req.currentUser.orgId, req.body));
  } catch (err) {
    console.error('[POST /jobs/jsearch/scheduler/start] error:', err);
    res.status(502).json({ detail: `Scheduler start error: ${err.message}` });
  }
});

router.post('/jsearch/scheduler/stop', authenticate, async (req, res) => {
  try {
    res.json(await stopJsearchSchedulerForOrg(req.currentUser.orgId));
  } catch (err) {
    console.error('[POST /jobs/jsearch/scheduler/stop] error:', err);
    res.status(502).json({ detail: `Scheduler stop error: ${err.message}` });
  }
});

export default router;